File: cs/2_7_flyweight.cs
/*
* при розробці прикладів використано матеріали
* https://refactoring.guru/ru/design-patterns/flyweight/csharp/example
*/
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Structural.Flyweight
{
    //вихідний стан

    public class Car
    {
        public string Owner { get; set; }
        public string Number { get; set; }
        public string Company { get; set; }
        public string Model { get; set; }
        public string Color { get; set; }

        public override string ToString()
        {
            return JsonSerializer.Serialize(this);
        }
    }
    /*
    * Пристосуванець зберігає спільну частину стану (також відому як внутрішній стан), яка належеть декільком реальним обєктам. 
    * Пристосуванець приймє решту стану (зовнішній стан, унікальний для кожного обєкту) через параемтри методу
    * Також пристосуванець  може повернути оригінальний об'єкт 
    */


    //внутрішній стан
    public class CarSharedState
    {
        public string Company { get; set; }
        public string Model { get; set; }

        public CarSharedState(Car car)
        {
            this.Company = car.Company;
            this.Model = car.Model;
        }

        public override string ToString()
        {
            return JsonSerializer.Serialize(this);
        }
    }

    //зовнішній стан
    public class CarUniqueState
    {
        public string Owner { get; set; }
        public string Number { get; set; }
        public string Color { get; set; }

        public CarUniqueState(Car car)
        {
            this.Owner = car.Owner;
            this.Number = car.Number;
            this.Color = car.Color;
        }

        public override string ToString()
        {
            return JsonSerializer.Serialize(this);
        }
    }

    public class Flyweight
    {
        private CarSharedState _sharedState;
        private CarUniqueState _uniqueState;

        public Flyweight(Car car)
        {
            _sharedState = new CarSharedState(car);
            _uniqueState = new CarUniqueState(car);
        }

        public Flyweight(CarSharedState sharedState)
        {
            _sharedState = sharedState;
        }

        public void SetUniqueState(Car uniqueState)
        {
            _uniqueState = new CarUniqueState(uniqueState);
        }

        public override string ToString()
        {
            string s = JsonSerializer.Serialize(this._sharedState);
            string u = JsonSerializer.Serialize(this._uniqueState);
            return $"Flyweight: shared {s} and unique {u} state";
        }

        public Car GetStandartObject()
        {
            return new Car()
            {
                Company = _sharedState.Company,
                Model = _sharedState.Model,
                Owner = _uniqueState.Owner,
                Number = _uniqueState.Number,
                Color = _uniqueState.Color
            };
        }

        /*
        Також пристосуванець може мати  властивості з гетерами та сетерами для зовнішнього стану 
        і властивості з гетерами для внутрішнього стану. Сетери властивостей внутрішнього стану КАТЕГОРИЧНО ЗАБОРОНЕНІ
        */
        public string Owner
        {
            get
            {
                return _uniqueState.Owner;
            }
            set
            {
                _uniqueState.Owner = value;
            }
        }
        public string Number
        {
            get
            {
                return _uniqueState.Number;
            }
            set
            {
                _uniqueState.Number = value;
            }
        }
        public string Color
        {
            get
            {
                return _uniqueState.Color;
            }
            set
            {
                _uniqueState.Color = value;
            }
        }

        public string Model
        {
            get
            {
                return _sharedState.Model;
            }
        }

        public string Company
        {
            get
            {
                return _sharedState.Company;
            }
        }
    }


    /*
    * Фабрика пристосуванців створює обєкти та керує ними.
    * Вона забезпечує правильний розподіл внутрішнього стану.
    * При створенні пристосуванця фабрика повертає існуючий спільний стан чи створює новий.
    */
    public class FlyweightFactory
    {
        private Dictionary<string, CarSharedState> sharedStates = new Dictionary<string, CarSharedState>();

        public FlyweightFactory(params Car[] args)
        {
            foreach (var elem in args)
            {
                sharedStates.Add(getKey(elem), new CarSharedState(elem));
            }
        }

        // Повертає текстовий геш для спільного стану 
        private string getKey(Car key)
        {
            return $"{key?.Company}_{key?.Model}";
        }

        // Повертає існуючий чи створює новий внутрішній стан пристосуванця
        public Flyweight GetFlyweight(Car car)
        {
            string key = getKey(car);
            // якщо нема жодного спільного стану з вказаним ключем, то його потрібно створити 
            if (!sharedStates.ContainsKey(key))
            {
                // Console.WriteLine("FlyweightFactory: Can't find a flyweight, creating new one.");
                sharedStates.Add(key, new CarSharedState(car));
            }
            else
            {
                // Console.WriteLine("FlyweightFactory: Reusing existing flyweight.");
            }
            // встановлюємо спільний сатн пристосуванця із закешованого списку
            CarSharedState sharedState = sharedStates.GetValueOrDefault(key);
            Flyweight flyweight = new Flyweight(sharedState);
            // встановлюємо унікальний стан пристосуваннця 
            flyweight.SetUniqueState(car);
            return flyweight;
        }

        public void listSharedStates()
        {
            Console.WriteLine($"\nFlyweightFactory: I have {sharedStates.Count} shared states:");
            foreach (var sharedState in sharedStates)
            {
                Console.WriteLine(sharedState);
            }
        }
    }
}

File: python/patterns/_2_7_flyweight.py
from dataclasses import dataclass
from typing import Optional, Self
import hashlib

# вихідний стан обєкту


@dataclass
class Car:
    owner: str = ""
    number: str = ""
    company: str = ""
    model: str = ""
    color: str = ""

#   Пристосуванець зберігає спільну частину стану (також відому як внутрішній стан), яка належеть декільком реальним обєктам.
#   Пристосуванець приймє решту стану (зовнішній стан, унікальний для кожного обєкту) через параемтри методу
#   Також пристосуванець  може повернути оригінальний об'єкт

# внутрішній стан


@dataclass
class CarSharedState:
    company: str = ""
    model: str = ""

    def __init__(self, car: Car) -> None:
        self.company = car.company
        self.model = car.model


# //зовнішній стан


@dataclass
class CarUniqueState:
    owner: str = ""
    number: str = ""
    color: str = ""

    def __init__(self, car: Car) -> None:
        self.owner = car.owner
        self.number = car.number
        self.color = car.color


class Flyweight:
    def __init__(self, car: Optional[Car] = None) -> None:
        if car is None:
            return
        self._shared_state: CarSharedState = CarSharedState(car)
        self._unique_state: CarUniqueState = CarUniqueState(car)

    def set_unique_state(self, car_unique_state: CarUniqueState) -> Self:
        self._unique_state = car_unique_state
        return self

    def set_shared_state(self, car_shared_state: CarSharedState) -> Self:
        self._shared_state = car_shared_state
        return self

    def __repr__(self) -> str:
        return f"Flyweight: shared {self._shared_state} and unique {self._unique_state} state"

    def to_object(self) -> Car:
        return Car(
            company=self._shared_state.company,
            model=self._shared_state.model,
            owner=self._unique_state.owner,
            number=self._unique_state.number,
            color=self._unique_state.color
        )
# Також пристосуванець може мати  властивості з гетерами та сетерами для зовнішнього стану
# і властивості з гетерами для внутрішнього стану. Сетери властивостей внутрішнього стану КАТЕГОРИЧНО ЗАБОРОНЕНІ

    @property
    def company(self) -> str:
        return self._shared_state.company

    @property
    def model(self) -> str:
        return self._shared_state.model

    @property
    def owner(self) -> str:
        return self._unique_state.owner

    @owner.setter
    def owner(self, value: str) -> None:
        self._unique_state.owner = value

    @property
    def number(self) -> str:
        return self._unique_state.number

    @number.setter
    def number(self, value: str) -> None:
        self._unique_state.number = value

    @property
    def color(self) -> str:
        return self._unique_state.color

    @color.setter
    def color(self, value: str) -> None:
        self._unique_state.color = value


#     Фабрика пристосуванців створює обєкти та керує ними.
#  Вона забезпечує правильний розподіл внутрішнього стану.
#     При створенні пристосуванця фабрика повертає існуючий спільний стан чи створює новий.
class FlyweightFactoryMethod:
    def __init__(self, *args: Car) -> None:
        self._shared_states: dict[str, CarSharedState] = {}
        for car in args:
            self._shared_states[self.get_key(car)] = CarSharedState(car)

    # Повертає текстовий геш для спільного стану
    def get_key(self, car: Car) -> str:
        raw = f"{car.company.strip().lower()}_{car.model.strip().lower()}"
        return hashlib.md5(raw.encode("utf-8")).hexdigest()

    # Повертає існуючий чи створює новий внутрішній стан пристосуванця
    def to_flyweight(self, car: Car) -> Flyweight:
        key: str = self.get_key(car)
        if not key in self._shared_states:
            self._shared_states[key] = CarSharedState(car)
        return Flyweight()\
            .set_shared_state(self._shared_states[key])\
            .set_unique_state(CarUniqueState(car))

    def print_shared_states(self) -> None:
        print(
            f"FlyweightFactory: I have {len(self._shared_states)} shared states:")
        for state in self._shared_states.values():
            print(state)
